apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+
+namespace Google.Cloud.Spanner
+{
+    internal static class ClientPool
+    {
+        private static readonly ConcurrentDictionary<ClientPoolKey, ClientPoolEntry> s_clientEntryPool =
+            new ConcurrentDictionary<ClientPoolKey, ClientPoolEntry>();
+
+        public static int Timeout { get; set; }
+
+        public static Task<SpannerClient> AcquireClientAsync()
+        {
+            return AcquireClientAsync(null, null);
+        }
+
+        public static async Task<SpannerClient> AcquireClientAsync(ITokenAccess credentials, ServiceEndpoint endpoint)
+        {
+            var key = new ClientPoolKey {
+                Credential = credentials ?? ApplicationDefault.Instance,
+                Endpoint = endpoint ?? SpannerClient.DefaultEndpoint
+            };
+            var poolEntry = s_clientEntryPool.GetOrAdd(key, k => new ClientPoolEntry(key));
+            var result = await poolEntry.AcquireClientFromEntryAsync().ConfigureAwait(false);
+            Logger.LogPerformanceCounter("SpannerClient.Count", () => s_clientEntryPool.Count);
+            return result;
+        }
+
+        /// <summary>
+        ///     Callers should be careful not to call AcquireClient while CloseAll is still running.
+        ///     It is possible to obtain a stale client.
+        /// </summary>
+        /// <returns></returns>
+        public static async Task CloseAllAsync()
+        {
+            Logger.Debug(() => "Shutting down all gRPC channels.");
+            await SpannerClient.ShutdownDefaultChannelsAsync().ConfigureAwait(false);
+            s_clientEntryPool.Clear();
+        }
+
+        private struct ClientPoolKey
+        {
+            public ITokenAccess Credential;
+            public ServiceEndpoint Endpoint;
+        }
+
+        private class ClientPoolEntry
+        {
+            private readonly ClientPoolKey _key;
+            private SpannerClient _client;
+
+            public ClientPoolEntry(ClientPoolKey key)
+            {
+                _key = key;
+            }
+
+            public async Task<SpannerClient> AcquireClientFromEntryAsync()
+            {
+                if (_client == null)
+                {
+                    Logger.Debug(() => "Creating a new SpannerClient.");
+                    if (_key.Credential != null && _key.Credential != ApplicationDefault.Instance)
+                        _client = await SpannerClient.CreateAsync(_key.Endpoint ?? SpannerClient.DefaultEndpoint,
+                                new SpannerSettings {
+                                    CallSettings = CallSettings.FromCallCredentials(_key.Credential.ToCallCredentials())
```

Comment:
It's still unclear to me why we're using this as *call* settings rather than *channel* settings.
(Doesn't this mean it's impossible to use the library with non-default credentials?)